pkg/lightspeed/funcs.go

```diff
@@ -0,0 +1,237 @@
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	uns "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+const (
+	// OpenStackLightspeedDefaultProvider - contains default name for the provider created in OLSConfig
+	// by openstack-operator.
+	OpenStackLightspeedDefaultProvider = "openstack-lightspeed-provider"
+
+	// OpenStackLightspeedOwnerIDLabel - name of a label that contains ID of OpenStackLightspeed instance
+	// that manages the OLSConfig.
+	OpenStackLightspeedOwnerIDLabel = "openstack.org/lightspeed-owner-id"
+
+	// OpenStackLightspeedVectorDBPath - path inside of the container image where the vector DB are
+	// located
+	OpenStackLightspeedVectorDBPath = "/rag/vector_db/os_product_docs"
+
+	// OLSConfigName - OLS forbids other name for OLSConfig instance than OLSConfigName
+	OLSConfigName = "cluster"
+)
+
+// GetOLSConfig returns OLSConfig if there is one present in the cluster.
+func GetOLSConfig(ctx context.Context, helper *common_helper.Helper) (uns.Unstructured, error) {
+	OLSConfigGVR := schema.GroupVersionResource{
+		Group:    "ols.openshift.io",
+		Version:  "v1alpha1",
+		Resource: "olsconfigs",
+	}
+
+	OLSConfigList := &uns.UnstructuredList{}
+	OLSConfigList.SetGroupVersionKind(OLSConfigGVR.GroupVersion().WithKind("OLSConfig"))
+	err := helper.GetClient().List(ctx, OLSConfigList)
+	if err != nil {
+		return uns.Unstructured{}, err
+	}
+
+	if OLSConfigList.Items != nil && len(OLSConfigList.Items) > 0 {
+		return OLSConfigList.Items[0], nil
+	}
+
+	return uns.Unstructured{}, k8s_errors.NewNotFound(
+		schema.GroupResource{Group: "ols.openshifg.io", Resource: "olsconfigs"},
+		"OLSConfig")
+}
+
+// IsOLSOperatorInstalled checks whether OLS Operator is already running in the cluster.
+func IsOLSOperatorInstalled(ctx context.Context, helper *common_helper.Helper) (bool, error) {
+	csvGVR := schema.GroupVersionResource{
+		Group:    "operators.coreos.com",
+		Version:  "v1alpha1",
+		Resource: "clusterserviceversions",
+	}
+
+	csvList := &uns.UnstructuredList{}
+	csvList.SetGroupVersionKind(csvGVR.GroupVersion().WithKind("clusterserviceversion"))
+	err := helper.GetClient().List(ctx, csvList)
+	if err != nil {
+		return false, err
+	}
+
+	for _, csv := range csvList.Items {
+		if strings.HasPrefix(csv.GetName(), "lightspeed-operator") {
+			return true, nil
+		}
+	}
+
+	return false, nil
+}
+
+// PatchOLSConfig patches OLSConfig with information from OpenStackLightspeed instance.
+func PatchOLSConfig(
+	olsConfig *uns.Unstructured,
+	instance *lightspeedv1beta1.OpenStackLightspeed,
+	helper *common_helper.Helper,
+) error {
+	// 1. Patch the Providers section
+	providersPatch := []interface{}{
+		map[string]interface{}{
+			"credentialsSecretRef": map[string]interface{}{
+				"name": StringPtrValue(instance.Spec.LLMCredentials),
```

Comment:
is it good to use StringPtrValue and possibly pass an empty string as the name? wouldn't it be better to check all the required files for being nil and return an explicite error? or is the intend to get the error from the api server when try to apply the obj with the missing required params?

Reply:
I've added the checks for nil for the Optional parameters to ensure that we do not pass `""` string to those if they are not set. For the other parameters, I think we should be good because they are marked as required and if they are not set they would be caught during the validation of the incoming `OpenStackControlPlane`. I'm marking as resolved. Is this resolution ok?